Resources/js/router.ts

```diff
@@ -0,0 +1,348 @@
+export interface RouteDefaults {
+  [index: string]: string;
+}
+
+export interface RouteRequirements {
+  [index: string]: string;
+}
+
+export interface RouteParams {
+  [index: string]: any;
+}
+
+export interface QueryParamAddFunction {
+  (prefix: string, params: any): void;
+}
+
+export interface Route {
+  tokens: ([string, string] | [string, string, string] | [string, string, string, string] | [string, string, string, string, boolean])[];
+  defaults: undefined[] | RouteDefaults;
+  requirements: undefined[] | RouteRequirements;
```

Comment:
Because of JSON serialization from PHP, an empty array is serialized to `[]` but a non-empty array is serialized to `{ 'default1': '...', 'default2: '...' }`